crates/sp-lightclient/src/lib.rs

```diff
@@ -79,22 +81,140 @@ pub trait Storage<Header: HeaderT> {
+                }
+                return Ok(());
+            }
+            Some(number) => number,
+        };
+
+        // we want to finalize the header from the current finalized header until the k-depth number of the best.
+        // 1. An ideal scenario, the current finalized head is one number less than number to be finalized but
+        // 2. if there was a re-org to longer chain when new header was imported, we do not want to miss
+        //    pruning fork headers between current and to be finalized number. So we go number by number and prune fork headers.
+        // 3. if there was a re-rg to a shorter chain when the new header was imported, then as per k-depth
+        //    we would have finalized more headers and, unfortunately, would have pruned any valid forks
+        //    There is not much we can do at this point then to just move the finalized head back and proceed.
+        match number_to_finalize.cmp(&current_finalized_number) {
+            // move the finalized head back to the expected head and return
+            Ordering::Less => {
+                let head_hash = self.heads_at_number(number_to_finalize)[0];
+                self.finalize_header(head_hash);
+            }
+            // nothing to do as we finalized the header already.
+            Ordering::Equal => (),
+            // move step by step from current to expected and prune all fork headers
+            Ordering::Greater => {
+                let mut current_finalized_number = current_finalized_number;
+
+                while current_finalized_number < number_to_finalize {
+                    current_finalized_number = current_finalized_number
+                        .checked_add(&One::one())
+                        .expect("should not overflow");
+
+                    // find the ancestor of the best header at k-depth
+                    let ancestor_header = self
+                        .ancestor_of_header_at_depth(current_finalized_number, &best_header)
+                        .expect("ancestor must exist at this point in time");
+
+                    // finalize the ancestor
+                    let mut number = *ancestor_header.header.number();
+                    let mut pruned_headers = self.finalize_header(ancestor_header.header.hash());
+
+                    // start pruning descendants of the forks at the finalized number
+                    while !pruned_headers.is_empty() {
+                        number = number
+                            .checked_add(&One::one())
+                            .expect("should not overflow");
+                        pruned_headers = self.prune_headers_with_parents_at_number(
+                            number,
+                            pruned_headers
+                                .into_iter()
+                                .map(|header| header.header.hash())
+                                .collect(),
+                        );
+                    }
```

Comment:
This API is really hard to read and comprehend. The harder it is, the higher probability that people will give up trying and may miss implementation bug. Both `prune_headers_with_parents_at_number` name and comment are non-intuitive to me. Can you describe with plain English what you're trying to achieve here?

Reply:
yeah. will do!